source/extensions/common/wasm/context.cc

```diff
@@ -202,6 +204,53 @@ void Context::onResolveDns(uint32_t token, Envoy::Network::DnsResolver::Resoluti
   wasm()->on_resolve_dns_(this, id_, token, s);
 }
 
+void Context::onStat(Envoy::Stats::MetricSnapshot& snapshot) {
+  // {name}.{delta/value}|{c/g}/n
+  proxy_wasm::DeferAfterCallActions actions(this);
+  uint32_t s = 0;
+  for (const auto& counter : snapshot.counters()) {
+    if (counter.counter_.get().used()) {
+      s += absl::StrCat(counter.counter_.get().name(), ".", counter.delta_, "|c/n").size() + 1;
+    }
+  }
+
+  for (const auto& gauge : snapshot.gauges()) {
+    if (gauge.get().used()) {
+      s += absl::StrCat(gauge.get().name(), ".", gauge.get().value(), "|g/n").size() + 1;
+    }
+  }
+  s += sizeof(uint32_t);
+
+  auto buffer = std::unique_ptr<char[]>(new char[s]);
+  char* b = buffer.get();
+
+  uint32_t n = snapshot.counters().size() + snapshot.gauges().size();
+  memcpy(b, &n, sizeof(uint32_t));
+  b += sizeof(uint32_t);
+
+  for (const auto& counter : snapshot.counters()) {
+    if (counter.counter_.get().used()) {
+      absl::string_view flush_counter =
+          absl::StrCat(counter.counter_.get().name(), ".", counter.delta_, "|c/n");
+      memcpy(b, flush_counter.data(), flush_counter.size());
```

Comment:
likewise, see if you can copy the bits without constructing the string for efficiency.  Note: you could either calculate the the size of the counter value string or cache the converted string.

Reply:
Resolving this comment since we have moved to a new stats marshalling approach in the revision